### ISSUE TITLE
Export for arrangement group

### ISSUE THREAD

**Issue by codemanyak** [fesch/Structorizer.Desktop#828]
Code export of a diagram may involve all directly or indirectly referenced diagrams (includables, subroutines).

But it would also be nice to be able to export an arrangement group with diagrams not referencing one another into a single file, e.g. as kind of library. This function might be added to the context menu of the Arranger Index. This would ensure that all members of the group be exported (indepently of whether called subroutines get involved or not).

**Comment by codemanyak** [comment]
It should not be forgotten to generate e.g. USES or import clauses if diagrams of the exported group refer to diagrams from other present groups and these are not to be involved (switched off export preference!)

**Comment by codemanyak** [comment]
The export of an arrangement group turned out to be a very complex task requiring a case analysis, in particular as the possible solutions may depend on the target language. The same problems arise on arrangement export in batch mode (#696, #836) but there are some more specific aspects.
We will discuss the common problems here and the more specific aspects in #836, therefore.

At first glance we might handle a group just as a bunch of diagrams. But this doesn't make sense of course as there may be call and include dependencies among the diagrams. These dependencies will have to be considered on code export since many languages require subroutines to be declared or defined before others may reference them (like C, Pascal, or shell scripts usually do).
A dependency analysis will reveal a dependency hierarchy that might be imagined as a forest of trees with the diagrams not included or called by any other diagram of the group as tree roots or entry points. In some cases even Includable diagrams may be such a tree root. Unfortunately, the trees of the forest are not necessarily disjoint, i.e. they may share some dependencies.
What cases are to be distinguished now?
1. There is a single tree with a main diagram as root - this can be treated as a classical program or application with auxiliary subroutines, possibly sharing some Includables. In object-oriented languages like Java it would become a class with a single main method and some private methods. (Whether these private methods are to be static or instance methods is of secondary interest. As emerging from Structorizer where they work like global functions, it will work to export them as static methods and the included resources as static attributes, though to export them as instance members might be more desirable in many cases.)
1. There is no main diagram but one ore more subroutine diagrams as tree roots. Such a group is certainly to be treated as a library, module (Oberon), or unit (Pascal). In Java they may form a common class without main method but some public static methods as entry points The good question is whether all of the routines or only the tree roots are to be exposed as entry points (e.g. to be declared in a C header file or as public in a class, or to be marked as exportable in an Oberon module). This question does not vanish if there is only a single tree root: Should all non-tree nodes be hidden? We can hardly decide this in general. Maybe a new export option may specify it in future.
1. There is exactly one tree with a main diagram as root and at least one more tree with an independent subroutine diagram as root. In this case, we will have to identify the set of subroutine and includable diagrams required by several trees. Now it depends on the target language what to do: In a procedural language, this shared subset should certainly form a kind of library, module, or unit, respectively, whereas the tree roots are to become a program/application and further libraries/modules/units not containing the shared routines but importing the common library/module/unit. In object-oriented languages, however, all generated methods may still go into a single class where just the entry points are marked as public (or see previous case).
1. More than one tree is rooted by a main diagram. This construction may make sense in Structorizer but is usually not allowed e.g. in a C project, which must not contain more than one main function. (Neither it makes sense to produce e.g. a single shell script containing several programs or a Java class with more than one main method.) So we will have to partition the group. Again, there might be shared subroutines or includables. The solution is analogous to the one for procedural languages of the previous case: We will have to produce a library/module/unit/class of all the routines shared by more than one tree and to export the difference sets of the found trees to separate programs/applications (for the main-rooted trees) or libraries/modules/units/classes (for the subroutine-rooted trees). Those tree export results requiring any of the routines in the common library must import/use/include it.

The correct result in cases 3 and 4 would be several source files (and possibly additional header files in case of C and C++). This might get tricky with respect to the file existence check in previous file chooser dialog (and with batch export), as well. So we might possibly adopt the recent practice of the batch export to write several file contents into the one previously selected target file and just insert scissor lines where to split the text into separate files.

So what we need as a result of the analysis phase is:
- All tree roots, i.e., all diagrams not required by any other. (Main diagrams are automatically tree roots.)
- The union of the pairwise intersections of all trees (as content of the common library).
- The difference sets of all the dependency trees with the previously mentioned shared set (this could be filtered on exporting the tree, though).
- The flag for every entry point whether its original tree had shared a diagram with the library (would be nice to have this information before starting to export the respective tree, otherwise requires a later insertion in the code).

Final remark: The above considerations should be copied (or "translated") to the User Guide to explain the users what decisions had to be made and what results are to be expected.

**Comment by codemanyak** [comment]
Version 3.30-08 revealed a flaw in group export: Unreferenced subroutines could be missing in the gerenated file in case the group contains exactly one main diagram (as if this main allone had been exported with export option "Involve called subroutines" enabled). Will be fixed in version 3.30-09.
